### PR TITLE
Implement ServerConfig admin REST API

### PR DESCRIPTION
**Pull request by krisis**
Returns a valid config.json of the setup. In case of distributed
setup, it checks if quorum or more number of nodes have the same
config.json.

## Description

Fetches config.json from all disks and returns a valid config.json if present.

## Motivation and Context

For users to be able to read config.json of a minio instance without having to log onto the corresponding server and fetch the config.json.
## How Has This Been Tested?

Manually tested using pkg/madmin API also part of this PR.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by krisis** [comment]
TBD:
- Type to represent config.json in pkg/madmin. For now it is the same as that used in minio server.

Rationale for getPeerConfig to use ``interface{}`` to ``Unmarshal`` the json bytes
- type-specific un-marshalling need not be unique for a given value of the type.
- un-marshalling to ``interface{}`` type invokes the rules as described [here](https://golang.org/pkg/encoding/json/#Unmarshal). This is guaranteed to be canonical.

**Comment by harshavardhana** [comment]
> Rationale for getPeerConfig to use interface{} to Unmarshal the json bytes
> 
> type-specific un-marshalling need not be unique for a given value of the type.
> un-marshalling to interface{} type invokes the rules as described here. This is guaranteed to be canonical.

Not sure i understand this can you elaborate? @krisis

**Comment by krisis** [comment]
@harshavardhana there are 2 reasons to use `interface{}` as the map key and not `ServerConfigV13`,
1. `ServerConfigV13` is not a valid key type - fails to compile.
2. `json` Unmarshal/Marshal for a given type may change with newer versions of the package.

**Comment by harshavardhana** [comment]
> ServerConfigV13 is not a valid key type - fails to compile.

Why do you need `ServerConfigV13` ? you can use what is available `serverConfigV13` 

> json Unmarshal/Marshal for a given type may change with newer versions of the package.

That is why we need to keep the migration code - interface{} doesn't add any value - all the more it will mask issues.

**Comment by krisis** [comment]
> >   ServerConfigV13 is not a valid key type - fails to compile.
> > Why do you need ServerConfigV13 ? you can use what is available serverConfigV13

`serverConfigV13` is the same as `ServerConfigV13`. We are capitalizing the type name to export it in pkg/madmin. So, both the types are not valid key type.

File: cmd/admin-handlers.go [hunk 1]
```diff
@@ -672,3 +672,23 @@ func (adminAPI adminAPIHandlers) HealFormatHandler(w http.ResponseWriter, r *htt
 	// Return 200 on success.
 	writeSuccessResponseHeadersOnly(w)
 }
+
```
Comment (harshavardhana) [inline cmd/admin-handlers.go:675]: Can you write a comment here.

File: cmd/admin-handlers.go [hunk 2]
```diff
@@ -672,3 +672,23 @@ func (adminAPI adminAPIHandlers) HealFormatHandler(w http.ResponseWriter, r *htt
...
+func (adminAPI adminAPIHandlers) GetConfigHandler(w http.ResponseWriter, r *http.Request) {
+	// Validate request signature.
+	adminAPIErr := checkRequestAuthType(r, "", "", "")
+	if adminAPIErr != ErrNone {
+		writeErrorResponse(w, adminAPIErr, r.URL)
+		return
+	}
```
Comment (harshavardhana) [inline cmd/admin-handlers.go:685]: i guess we should avoid meddling with server config if object layer is not initialized, safer that way.

File: cmd/admin-rpc-client.go [hunk 3]
```diff
@@ -112,6 +116,23 @@ func (rc remoteAdminClient) Uptime() (time.Duration, error) {
 	return reply.Uptime, nil
 }
 
+func (lc localAdminClient) GetConfig() ([]byte, error) {
```
Comment (harshavardhana) [inline cmd/admin-rpc-client.go]: Comment.

File: cmd/admin-rpc-client.go [hunk 4]
```diff
@@ -112,6 +116,23 @@ func (rc remoteAdminClient) Uptime() (time.Duration, error) {
...
+	if serverConfig == nil {
+		return nil, errors.New("config not present")
+	}
+
+	return json.Marshal(serverConfig)
+}
+
```
Comment (harshavardhana) [inline cmd/admin-rpc-client.go:127]: Comment.

File: cmd/admin-rpc-client.go [hunk 5]
```diff
@@ -336,3 +357,116 @@ func getPeerUptimes(peers adminPeers) (time.Duration, error) {
 
 	return latestUptime, nil
 }
+
+func getPeerConfig(peers adminPeers) ([]byte, error) {
```
Comment (harshavardhana) [inline cmd/admin-rpc-client.go]: A comment.

File: cmd/admin-rpc-client.go [hunk 6]
```diff
@@ -336,3 +357,116 @@ func getPeerUptimes(peers adminPeers) (time.Duration, error) {
...
+	if err != nil {
+		errorIf(err, "Unable to find a valid server config")
+		return nil, traceError(err)
+	}
+
+	// Return the config.json that was present quorum or more
+	// number of disks.
+	return json.Marshal(configJSON)
+}
+
```
Comment (harshavardhana) [inline cmd/admin-rpc-client.go:411]: Comment.

**Review by harshavardhana — CHANGES_REQUESTED** [review]

File: cmd/admin-handlers.go (hunk 1 above)
Comment (krisis) [reply cmd/admin-handlers.go:675]: Done

File: cmd/admin-handlers.go (hunk 2 above)
Comment (krisis) [reply cmd/admin-handlers.go:685]: Ok

File: cmd/admin-rpc-client.go (hunk 4 above)
Comment (krisis) [reply cmd/admin-rpc-client.go:127]: Done.

File: cmd/admin-rpc-client.go (hunk 6 above)
Comment (krisis) [reply cmd/admin-rpc-client.go:411]: Done

File: cmd/admin-rpc-client.go [hunk 7]
```diff
@@ -336,3 +359,120 @@ func getPeerUptimes(peers adminPeers) (time.Duration, error) {
...
+	// serverConfigs == [c1, c2, c1, c1]
+	// configCounter == [3, 1, 0, 0]
+	// c1, c2 are the only distinct values that appear.  c1 is
+	// identified by 0, the lowest index it appears in and c2 is
+	// identified by 1. So, we need to find the number of times
+	// each of these distinct values occur.
+
+	// For each of the serverConfig ...
+	for i := range serverConfigs {
+		// Skip nodes were getConfig failed.
```
Comment (donatello) [inline cmd/admin-rpc-client.go]: `s/were/where/`

File: cmd/admin-rpc-client.go [hunk 8]
```diff
@@ -336,3 +359,120 @@ func getPeerUptimes(peers adminPeers) (time.Duration, error) {
...
+				// is the lower index.
+				configCounter[j]++
+				break
+			} else if j == i {
+				// serverConfigs[i] is equal to no
+				// other value seen before. i.e, it is
+				// unique so far.
+				configCounter[i] = 1
+				break
+			} // else invariance is violated.
```
Comment (donatello) [inline cmd/admin-rpc-client.go]: What is the invariance? It doesn't seem to be precisely stated anywhere.

The invariance is after `i-`th iteration, `sum of configCounter slice = i - errors seen upto i-th iteration`. Hopefully there's a nicer way to state it.

File: cmd/admin-rpc-client.go [hunk 9]
```diff
@@ -336,3 +359,120 @@ func getPeerUptimes(peers adminPeers) (time.Duration, error) {
...
+	for i := range serverConfigs {
+		// Skip nodes were getConfig failed.
+		if errs[i] != nil {
+			continue
+		}
+		// Find the already seen copy it is equal to.
+		for j := 0; j <= i; j++ {
+			if j < i && configCounter[j] == 0 {
+				// serverConfigs[j] is known to be equal to a
+				// value that was already seen. See e.g above
```
Comment (donatello) [inline cmd/admin-rpc-client.go]: `s/e.g/example/`

File: cmd/admin-rpc-client.go [hunk 10]
```diff
@@ -336,3 +359,120 @@ func getPeerUptimes(peers adminPeers) (time.Duration, error) {
...
+			} else if j < i && reflect.DeepEqual(serverConfigs[i], serverConfigs[j]) {
+				// serverConfigs[i] is equal to
+				// serverConfigs[j], update
+				// serverConfigs[j]'s counter since it
+				// is the lower index.
+				configCounter[j]++
+				break
+			} else if j == i {
+				// serverConfigs[i] is equal to no
+				// other value seen before. i.e, it is
```
Comment (donatello) [inline cmd/admin-rpc-client.go]: `s/i.e, it/It`

File: cmd/admin-rpc-client.go [hunk 11]
```diff
@@ -336,3 +359,120 @@ func getPeerUptimes(peers adminPeers) (time.Duration, error) {
...
+// quorum or more number of servers.
+func getValidServerConfig(serverConfigs []serverConfigV13, errs []error) (serverConfigV13, error) {
+	// majority-based quorum
+	quorum := len(serverConfigs)/2 + 1
+
+	// Count the number of disks a config.json was found in.
+	configCounter := make([]int, len(serverConfigs))
+
+	// We group equal serverConfigs by the lowest index of the
+	// same value.  e.g, let us take the following serverConfigs
```
Comment (donatello) [inline cmd/admin-rpc-client.go]: `s/value. e.g,/value; e.g./`

**Review by donatello — APPROVED** [review]
LGTM

**Review by harshavardhana — APPROVED** [review]

**Pull request merged**